p2p/http/libp2phttp.go

```diff
@@ -0,0 +1,744 @@
+	})
+	<-h.httpTransport.waitingForListeners
+	return h.httpTransport.listenAddrs
+}
+
+var ErrNoListeners = errors.New("nothing to listen on")
+
+// Serve starts the HTTP transport listeners. Always returns a non-nil error.
+// If there are no listeners, returns ErrNoListeners.
+func (h *HTTPHost) Serve() error {
+	// assert that each addr contains a /http component
+	for _, addr := range h.ListenAddrs {
+		_, isHTTP := normalizeHTTPMultiaddr(addr)
+		if !isHTTP {
+			return fmt.Errorf("address %s does not contain a /http or /https component", addr)
+		}
+	}
+
+	h.ServeMux.Handle("/.well-known/libp2p", &h.wk)
+
+	h.createHTTPTransport.Do(func() {
+		h.httpTransport = newHTTPTransport()
+	})
+
+	closedWaitingForListeners := false
+	defer func() {
+		if !closedWaitingForListeners {
+			close(h.httpTransport.waitingForListeners)
+		}
+	}()
+
+	if len(h.ListenAddrs) == 0 && h.StreamHost == nil {
+		return ErrNoListeners
+	}
+
+	h.httpTransport.listeners = make([]net.Listener, 0, len(h.ListenAddrs)+1) // +1 for stream host
+
+	streamHostAddrsCount := 0
+	if h.StreamHost != nil {
+		streamHostAddrsCount = len(h.StreamHost.Addrs())
+	}
+	h.httpTransport.listenAddrs = make([]ma.Multiaddr, 0, len(h.ListenAddrs)+streamHostAddrsCount)
+
+	errCh := make(chan error)
+
+	if h.StreamHost != nil {
+		listener, err := streamHostListen(h.StreamHost)
+		if err != nil {
+			return err
+		}
+		h.httpTransport.listeners = append(h.httpTransport.listeners, listener)
+		h.httpTransport.listenAddrs = append(h.httpTransport.listenAddrs, h.StreamHost.Addrs()...)
+
+		go func() {
+			errCh <- http.Serve(listener, &h.ServeMux)
+		}()
+	}
+
+	closeAllListeners := func() {
+		for _, l := range h.httpTransport.listeners {
+			l.Close()
+		}
+	}
+
+	err := func() error {
+		for _, addr := range h.ListenAddrs {
+			parsedAddr := parseMultiaddr(addr)
+			// resolve the host
+			ipaddr, err := net.ResolveIPAddr("ip", parsedAddr.host)
+			if err != nil {
+				return err
+			}
+
+			host := ipaddr.String()
+			l, err := net.Listen("tcp", host+":"+parsedAddr.port)
+			if err != nil {
+				return err
+			}
+			h.httpTransport.listeners = append(h.httpTransport.listeners, l)
+
+			// get resolved port
+			_, port, err := net.SplitHostPort(l.Addr().String())
+			if err != nil {
+				return err
+			}
+
+			var listenAddr ma.Multiaddr
+			if parsedAddr.useHTTPS && parsedAddr.sni != "" && parsedAddr.sni != host {
+				listenAddr = ma.StringCast(fmt.Sprintf("/ip4/%s/tcp/%s/tls/sni/%s/http", host, port, parsedAddr.sni))
+			} else {
+				scheme := "http"
+				if parsedAddr.useHTTPS {
+					scheme = "https"
+				}
+				listenAddr = ma.StringCast(fmt.Sprintf("/ip4/%s/tcp/%s/%s", host, port, scheme))
+
+			}
+
+			if parsedAddr.useHTTPS {
+				go func() {
+					srv := http.Server{
+						Handler:   &h.ServeMux,
+						TLSConfig: h.TLSConfig,
+					}
+					errCh <- srv.ServeTLS(l, "", "")
+				}()
+				h.httpTransport.listenAddrs = append(h.httpTransport.listenAddrs, listenAddr)
```

Comment:
This causes a race with `Addrs() `

If `Serve` is called in a separate goroutine, then there needs to be some synchronization to know when `Addrs()` cannot be called. Or, `h.httpTransport.listenAddrs` needs to be set in some function that is called before calling `Serve`. This worked previously because the constructor served that purpose.

Seems that a constructor is needed. Or, maybe a `Start` function that does this setup and calls `Serve` in a separate goroutine and returns.